### PR TITLE
fix: connector crashed when failed to connector to nats

### PR DESCRIPTION
**Pull request by luongngocminh**

### CODE REVIEW

**Comment by giangndm** [comment]
@luongngocminh I think it should handle error instead of ignore error. Current solution will lost package when NATS error.

The best solution is push event into queue (can be persistence on disk), only pop from queue if it successful sent to NATS, if NATS error it should reconnect and log error

**Review by giangndm — CHANGES_REQUESTED** [review]
We should implement mechanism for avoiding RPC handler being blocked, it simple put to a queue (persistence is the best, https://crates.io/crates/yaque), consume the queue the push to NATS, if it success consume the data.

File: servers/media-server/Cargo.toml [hunk 1]
```diff
@@ -41,5 +42,5 @@ webrtc = ["transport-webrtc"]
 rtmp = ["transport-rtmp"]
 sip = ["rsip", "transport-sip"]
 gateway = []
-connector = ["nats", "prost"]
+connector = ["nats", "prost", "queues"]
```
[the remark this answers is not in the document]
Comment (giangndm) [reply servers/media-server/Cargo.toml:45]: Or using https://crates.io/crates/yaque

File: servers/media-server/src/server/connector.rs [hunk 2]
```diff
@@ -57,22 +57,16 @@ where
...
-            }
-        }
+    let mut transporter: Box<dyn ConnectorTransporter<MediaEndpointLogRequest>> = match protocol.as_str() {
+        "nats" => Box::new(NatsTransporter::new(_opts.mq_uri.clone(), _opts.mq_channel.clone())),
         _ => {
             log::error!("Unsupported transporter");
             return Err("Unsupported transporter");
         }
     };
+    if let Err(e) = transporter.connect().await {
```
[the remark this answers is not in the document]
Comment (giangndm) [reply servers/media-server/src/server/connector.rs:67]: It seam to auto re-connect on transport

[the remark this answers is not in the document]
Comment (luongngocminh) [reply servers/media-server/src/server/connector.rs:67]: It will try to reconnect when there are logs waiting to be send, this shouldn't crash the server

**Comment by giangndm** [comment]
@luongngocminh Look good, please add some tests for ensure TransporterQueue working and fix some variable names. If the variable is used, shouldn't prefix it with "_"

**Review by giangndm — CHANGES_REQUESTED** [review]
@luongngocminh Look good, please add some tests for ensure TransporterQueue working and fix some variable names. If the variable is used, shouldn't prefix it with "_"

File: servers/media-server/src/server/connector/queue.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        async_std::task::spawn(async move {
+            let mut sender = sender;
+            sender.send(message_c.clone().encode_to_vec()).await.unwrap();
+        });
+
+        // Poll the TransporterQueue and assert the received data
+        queue.poll().await.unwrap();
+        assert!(rx.try_recv().is_err());
+
+        sleep(Duration::from_millis(1001)).await;
```
Comment (giangndm) [inline servers/media-server/src/server/connector/queue.rs:183]: Why we need sleep here?

File: servers/media-server/src/server/connector/queue.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    pub fn new(base: &str, transporter: Box<dyn ConnectorTransporter<M>>) -> Result<(Self, yaque::Sender), io::Error> {
+        if let Err(err) = yaque::recovery::recover(base) {
+            log::warn!("Error trying to recover queue, maybe first time: {:?}", err);
+        }
+        let (tx, rx) = yaque::channel(base)?;
+        Ok((Self { rx, transporter }, tx))
+    }
+
+    pub async fn poll(&mut self) -> Result<(), io::Error> {
+        log::debug!("Polling Nats transporter");
```
Comment (giangndm) [inline servers/media-server/src/server/connector/queue.rs:23]: in this context we don't know it is nats or whatever

File: servers/media-server/src/server/connector/queue.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        queue.poll().await.unwrap();
+        assert!(rx.try_recv().is_err());
+
+        sleep(Duration::from_millis(1001)).await;
+
+        // Poll the TransporterQueue again and assert the received data
+        queue.poll().await.unwrap();
+        let received_message = rx.try_recv().unwrap();
+
+        assert_eq!(message, received_message);
```
Comment (giangndm) [inline servers/media-server/src/server/connector/queue.rs:189]: need to check queue is empty

**Review by giangndm — CHANGES_REQUESTED** [review]

**Review by giangndm — APPROVED** [review]

**Pull request merged**